docs/implementation.adoc

```diff
@@ -0,0 +1,507 @@
+
+** `jenkins.scm.api.SCMSource` would represent a single Git repository.
+
+** `jenkins.scm.api.SCMNavigator` would represent a collection of Git repositories - most likely all on the same server
+
+* Subversion can have a fairly direct mapping of concepts analogous to the Git mapping above.
+Alternatively some oganizations will use a single Subversion repository to hold multiple projects under which each project would have the familiar `trunk/branches/tags` structure. If we consider this style of Subversion use then:
+
+** `jenkins.scm.api.SCMHead` would be mapped against a specific project and be one of: `trunk`, `branches/_name_` or `tags/_name_`.
+
+** `jenkins.scm.api.SCMRevision` would be mapped to the global Subversion revision number against which the specific `SCMHead` was last modified. Each `SCMRevision` is deterministic as if you ask to check out the trunk/branch.tag at the same revision you will always get the same contents.
+
+** `jenkins.scm.api.SCMSource` would represent a single project withi the Subversion repository.
+
+** `jenkins.scm.api.SCMNavigator` would represent the collection of projects in the root of the Subversion repository.
+
+* CVS, one mapping of the CVS concepts would be:
+
+** `jenkins.scm.api.SCMHead` would represent a branch in a module.
+
+** `jenkins.scm.api.SCMRevision` would represent a timestamp. Sadly the timestamp is not deterministic as an immediate checkout may miss some files that are still being committed to the repository.
++
+NOTE: An alternative deterministic `jenkins.scm.api.SCMRevision` would be a list of all the files and their individual revisions, but this would be prohibitive in terms of storage cost and is non-trivial for a user to replicate from their own workspace and consequently the timestamp would be preferred.
+Ignoring the problems of time sychronization in a distributed system, we could assume that any timestamp more than _X_ minutes in the past should be deterministic in order to allow the consuming plugins to make optimizations that are only possible with deterministic revisions.
+
+** `jenkins.scm.api.SCMSource` would represent an individual module on the CVS server.
+
+** `jenkins.scm.api.SCMNavigator` would represent the collection of modules available from a single CVS server.
+
+Implementers are free to map the concepts to their own SCM system as they see fit, but the recommendation is to try to keep close to the principles of mapping outlined in the above examples.
+
+The concepts we have covered so far determine how Jenkins plugins can drive interactions with the SCM system.
+While a Jenkins driven interaction with an SCM is sufficient for enabling advanced SCM functionality such as that provided in the https://wiki.jenkins-ci.org/display/JENKINS/Branch+API+Plugin[Branch API plugin], it does not lend to a good user experience as Jenkins would be required to continually poll the backing SCM to establish if there are any changes.
```

Comment:
This is the introduction, I am giving an overview of everything that needs to be implemented... then I start going through the different parts one by one, in the order that makes the most sense... perhaps I just need to rename this section as Overview and put a short introduction ahead of the Overview